load/gradual_load_orchestrator.go

```diff
@@ -0,0 +1,240 @@
+	generators []TxGenerator[T]
+	issuers    []Issuer[T]
+	tracker    Tracker[U]
+
+	log logging.Logger
+
+	maxObservedTPS uint64
+
+	observerGroup errgroup.Group
+	issuerGroup   *errgroup.Group
+
+	done chan struct{}
+
+	cancel context.CancelFunc
+
+	config GradualLoadOrchestratorConfig
+}
+
+func NewGradualLoadOrchestrator[T, U comparable](
+	generators []TxGenerator[T],
+	issuers []Issuer[T],
+	tracker Tracker[U],
+	log logging.Logger,
+	config GradualLoadOrchestratorConfig,
+) (*GradualLoadOrchestrator[T, U], error) {
+	if len(generators) != len(issuers) {
+		return nil, ErrMismatchedGeneratorsAndIssuers
+	}
+	return &GradualLoadOrchestrator[T, U]{
+		generators: generators,
+		issuers:    issuers,
+		tracker:    tracker,
+		log:        log,
+		config:     config,
+		done:       make(chan struct{}),
+	}, nil
+}
+
+func (o *GradualLoadOrchestrator[T, U]) Execute(ctx context.Context) error {
+	ctx, cancel := context.WithCancel(ctx)
+	o.cancel = cancel
+
+	// start a goroutine to confirm each issuer's transactions
+	for _, issuer := range o.issuers {
+		o.observerGroup.Go(func() error { return issuer.Listen(ctx) })
+	}
+
+	go o.run(ctx)
+
+	// wait for the test to finish
+	<-o.done
+
+	// stop the observers and issuers
+	o.cancel()
+
+	// block until both the observers and issuers have stopped
+	return errors.Join(o.issuerGroup.Wait(), o.observerGroup.Wait())
+}
+
+// run the gradual load test by continuously increasing the rate at which
+// transactions are sent
+//
+// run stops when the network can no longer make progress or if an issuer errors.
+func (o *GradualLoadOrchestrator[T, U]) run(ctx context.Context) {
+	defer close(o.done)
+
+	var (
+		prevConfirmed  = o.tracker.GetObservedConfirmed()
+		prevTime       = time.Now()
+		achievedMaxTPS bool
+		attempts       uint64
+	)
+
+	currTargetTPS := atomic.Uint64{}
+	currTargetTPS.Store(o.config.MinTPS)
+
+	issuerGroup, issuerCtx := errgroup.WithContext(ctx)
+	o.issuerGroup = issuerGroup
+
+	// start a goroutine to each issuer to continuously send transactions
+	// if an issuer errors, all other issuers will stop as well.
+	for i, issuer := range o.issuers {
+		o.issuerGroup.Go(func() error {
+			for {
+				select {
+				case <-issuerCtx.Done():
+					return issuerCtx.Err()
+				default:
+				}
+				currTime := time.Now()
+				txsPerIssuer := uint64(math.Ceil(float64(currTargetTPS.Load())/float64(len(o.issuers))) * o.config.TxMultiplier)
+				for range txsPerIssuer {
+					tx, err := o.generators[i].GenerateTx(issuerCtx)
+					if err != nil {
+						return err
+					}
+					if err := issuer.IssueTx(ctx, tx); err != nil {
+						return err
+					}
+				}
+				diff := time.Second - time.Since(currTime)
+				if diff > 0 {
+					time.Sleep(diff)
+				}
+			}
+		})
+	}
+
+	// blocks until either 1) the max TPS target has been reached, 2) we've
+	// maxed out the number of attempts, or 3) an issuer has errored.
+	for {
+		time.Sleep(o.config.SustainedTime)
```

Comment:
use `time.After` here so that you could quickly abort in case of context cancelation.